libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,206 @@
+		if (funType && funType->kind() == FunctionType::Kind::Internal)
+		{
+			processFunction(callable->resolveVirtual(*m_contract), _identifier.annotation().calledDirectly);
+
+			solAssert(m_currentNode.has_value(), "");
+		}
+	}
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(NewExpression const& _newExpression)
+{
+	if (ContractType const* contractType = dynamic_cast<ContractType const*>(_newExpression.typeName().annotation().type))
+		m_graph->createdContracts.emplace(&contractType->contractDefinition());
+
+	return true;
+}
+
+void FunctionCallGraphBuilder::endVisit(MemberAccess const& _memberAccess)
+{
+	auto functionType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type);
+	auto functionDef = dynamic_cast<FunctionDefinition const*>(_memberAccess.annotation().referencedDeclaration);
+	if (!functionType || !functionDef || functionType->kind() != FunctionType::Kind::Internal)
+		return;
+
+	// Super functions
+	if (*_memberAccess.annotation().requiredLookup == VirtualLookup::Super)
+	{
+		if (auto const* typeType = dynamic_cast<TypeType const*>(_memberAccess.expression().annotation().type))
+			if (auto const contractType = dynamic_cast<ContractType const*>(typeType->actualType()))
+			{
+				solAssert(contractType->isSuper(), "");
+				functionDef =
+					&functionDef->resolveVirtual(
+						*m_contract,
+						contractType->contractDefinition().superContract(*m_contract)
+					);
+			}
+	}
+	else
+		solAssert(*_memberAccess.annotation().requiredLookup == VirtualLookup::Static, "");
+
+	processFunction(*functionDef, _memberAccess.annotation().calledDirectly);
+	return;
+}
+
+void FunctionCallGraphBuilder::endVisit(ModifierInvocation const& _modifierInvocation)
+{
+	VirtualLookup const& requiredLookup = *_modifierInvocation.name().annotation().requiredLookup;
+
+	if (auto const* modifier = dynamic_cast<ModifierDefinition const*>(_modifierInvocation.name().annotation().referencedDeclaration))
+	{
+		if (requiredLookup == VirtualLookup::Virtual)
+			modifier = &modifier->resolveVirtual(*m_contract);
+		else
+			solAssert(requiredLookup == VirtualLookup::Static, "");
+
+		processFunction(*modifier, requiredLookup == VirtualLookup::Static);
+	}
+}
+
+void FunctionCallGraphBuilder::visitCallable(CallableDeclaration const* _callable, bool _directCall)
+{
+	solAssert(!m_graph->edges.count(_callable), "");
+
+	std::optional<Node> previousNode = m_currentNode;
+	m_currentNode = _callable;
+
+	if (previousNode.has_value() && _directCall)
+		add(*previousNode, _callable);
+	if (!_directCall)
+		add(*m_currentNode, m_currentDispatch);
+
+	_callable->accept(*this);
+
+	m_currentNode = previousNode;
+}
+
+bool FunctionCallGraphBuilder::add(Node _caller, Node _callee)
+{
+	return m_graph->edges[_caller].insert(_callee).second;
+}
+
+void FunctionCallGraphBuilder::processFunction(CallableDeclaration const& _callable, bool _calledDirectly)
+{
+	if (m_graph->edges.count(&_callable))
+		return;
```

Comment:
Wait a second... given what you said in the other thread, I'm now wondering:
This now *won't* create the dispatch edge below, whenever the function is already in the graph - aren't you missing dispatch edges then, if you first call a function directly and then afterwards call it indirectly?

Reply:
If the constructor calls a function directly, then it will visit here, but create no edge. But if that function then calls another function, there will be an edge from it to that function, i.e. it will already be in the graph. So later it won't end up in the dispatch. Right? Didn't look at it in detail, so maybe I'm wrong, but we should make sure it's ok...

Reply:
I think you're right, this is a detail I didn't notice. Though to fix this it should be enough to just move the check here below the dispatch add call, right?

Reply:
I don't think so - at least if it still actually makes a difference if ``visitCallable`` is called with ``_calledDirectly`` being true or false. Which relates to https://github.com/ethereum/solidity/pull/10332/files#r553951829. On the other hand I did argue that this could be rewritten, s.t. ``visitCallable`` *doesn't* need that argument in https://github.com/ethereum/solidity/pull/10332/files#r553385824 - if that works then: yes: moving that check would be a solution ;-).

In general it would be helpful if you clearly specified what ``processFunction`` and ``visitCallable`` is supposed to do exactly and what they are supposed to connect exactly - that way it'd be much easier to avoid oversights like this, resp. check if it does what it should.

Reply:
I fixed it by moving the check to `processFunction()`.

I'm also going to refactor these functions a bit because for me it's also not very clear what their responsibilities are. I mean, `visitCallable()` is a bit clearer because after my changes it just remember the old value if `m_currentNode` and visits the callable but I had some cases where `processFunction()` was doing almost what I needed but not quite and I ended up just doing `add()` directly + some extra instructions.